File: app/src/main/java/com/example/journal/LogInManager.java
package com.example.journal;

import android.os.Build;

import java.io.DataOutputStream;
import java.io.FileWriter;
import java.net.HttpCookie;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

public class LogInManager {
    private final String USER_AGENT = "Mozilla/5.0";
    private final String ROOT_DIRECTORY;

    private String sessionid;
    private String csrftoken;

    private String pupilUrl;
    private String postParameters;

    public LogInManager(String rootDirectory) {
        ROOT_DIRECTORY = rootDirectory;
    }

    private void setPostParameters(String un, String pw) {
        postParameters = "csrfmiddlewaretoken=";
        postParameters += csrftoken;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            postParameters += "&username=";
            try {
                postParameters += URLEncoder.encode(un, StandardCharsets.UTF_8.toString());
            } catch (Exception e) {
                postParameters += URLEncoder.encode(un);
            }

            postParameters += "&password=";
            try {
                postParameters += URLEncoder.encode(pw, StandardCharsets.UTF_8.toString());
            } catch (Exception e) {
                postParameters += URLEncoder.encode(pw);
            }
        }
        else {
            postParameters += "&username=";
            postParameters += URLEncoder.encode(un);
            postParameters += "&password=";
            postParameters += URLEncoder.encode(pw);
        }
    }

    public void writeLoginDataToFiles(String username, String realName) {
        try {
            FileWriter foutUsername = new FileWriter(ROOT_DIRECTORY + "/UserData/username.txt", false);
            foutUsername.write(username);
            foutUsername.flush();
            foutUsername.close();
        } catch (Exception e) {
            System.out.println("foutUsername error");
        }

        try {
            FileWriter foutSessionid = new FileWriter(ROOT_DIRECTORY + "/UserData/sessionid.txt", false);
            foutSessionid.write(sessionid);
            foutSessionid.flush();
            foutSessionid.close();
        } catch (Exception e) {
            System.out.println("foutSessionid error");
        }

        try {
            FileWriter foutUrl = new FileWriter(ROOT_DIRECTORY + "/UserData/pupilUrl.txt", false);
            foutUrl.write(pupilUrl);
            foutUrl.flush();
            foutUrl.close();
        } catch (Exception e) {
            System.out.println("foutUrl error");
        }

        try {
            FileWriter foutRealName = new FileWriter(ROOT_DIRECTORY + "/UserData/realName.txt", false);
            foutRealName.write(realName);
            foutRealName.flush();
            foutRealName.close();
        } catch (Exception e) {
            System.out.println("foutRealName error");
        }

        try {
            FileWriter foutStatus = new FileWriter(ROOT_DIRECTORY + "/UserData/status.txt", false);
            foutStatus.write("YES");
            foutStatus.flush();
            foutStatus.close();
        } catch (Exception e) {
            System.out.println("foutStatus error");
        }
    }

    public void takeCsrftoken() throws Exception {
        // open connection
        URL connectionUrl = new URL("https://schools.by/login");
        HttpURLConnection con = (HttpURLConnection)connectionUrl.openConnection();

        // set connection args
        con.setRequestMethod("GET");
        con.setRequestProperty("User-Agent", USER_AGENT);

        Map<String,List<String>> hf = con.getHeaderFields();
        for (Map.Entry<String, List<String>> ent : hf.entrySet()) {
            try {
                if (ent.getKey().equals("Set-Cookie")) {
                    List<HttpCookie> cookies = HttpCookie.parse(ent.getValue().get(0));
                    for (HttpCookie cookie : cookies) {
                        if (cookie.getName().equals("csrftoken")) {
                            csrftoken = cookie.getValue();
                            return;
                        }
                    }
                }
            } catch (NullPointerException npe) {
                continue;
            }
        }
    }

    public LoginState loggedIn(String username, String password) {
        try {
            takeCsrftoken();
        } catch (Exception e1) {
            try {
                takeCsrftoken();
            } catch (Exception e2) {
                return LoginState.ERROR_OCCURED;
            }
        }

        setPostParameters(username, password);

        URL connectionUrl;
        HttpURLConnection con;
        try {
            // open connection
            connectionUrl = new URL("https://schools.by/login");
            con = (HttpURLConnection) connectionUrl.openConnection();

            // configure connection
            con.setInstanceFollowRedirects(false);
            con.setUseCaches(false);

            con.setRequestMethod("POST");
            con.setRequestProperty("accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8");
            con.setRequestProperty("accept-encoding", "gzip, deflate, br");
            con.setRequestProperty("accept-language", "en-gb");
            con.setRequestProperty("Connection", "keep-alive");
            con.setRequestProperty("content-length", Integer.toString(postParameters.length()));
            con.setRequestProperty("content-type", "application/x-www-form-urlencoded");
            con.setRequestProperty("cookie", "csrftoken=" + csrftoken + ";"); //cookie
            con.setRequestProperty("origin", "https://schools.by");
            con.setRequestProperty("referer", "https://schools.by/login");
            con.setRequestProperty("user-agent", USER_AGENT);

            con.setDoOutput(true);
            con.setDoInput(true);
        } catch (Exception e) {
            return LoginState.ERROR_OCCURED;
        }

        // send data
        try {
            DataOutputStream stream = new DataOutputStream(con.getOutputStream());
            stream.writeBytes(postParameters);
            stream.flush();
            stream.close();
        } catch (Exception e) {
            return LoginState.ERROR_OCCURED;
        }

        try {
            con.getContent();
        } catch (Exception e) {
            return LoginState.ERROR_OCCURED;
        }


        // login check
        int responseCode = -1;
        try {
            responseCode = con.getResponseCode();
        } catch (Exception e) {
            return LoginState.ERROR_OCCURED;
        }
        while (responseCode == -1) {
            try {
                Thread.sleep(25);
            } catch (Exception e) {
                return LoginState.ERROR_OCCURED;
            }
        }
        if (responseCode == 200) {
            return LoginState.WRONG_PASSWORD;
        }

        // get csrftoken, sessionid, pupilUrl
        Map<String,List<String>> hf = con.getHeaderFields();
        for (Map.Entry<String, List<String>> ent : hf.entrySet()) {
            try {
                if (ent.getKey().equals("Set-Cookie")) {
                    for (String s : ent.getValue()) {
                        List<HttpCookie> cookies = HttpCookie.parse(s);
                        for (HttpCookie cookie : cookies) {
                            if (cookie.getName().equals("csrftoken")) {
                                csrftoken = cookie.getValue();
                            }
                            if (cookie.getName().equals("sessionid")) {
                                sessionid = cookie.getValue();
                            }
                        }
                    }
                }
                if (ent.getKey().equals("Location")) {
                    pupilUrl = ent.getValue().get(0);
                }
            } catch (NullPointerException npe) {
                continue;
            }
        }

        return LoginState.LOGGED_IN;
    }

    public String getSessionid() {
        return sessionid;
    }
    public String getPupilUrl() {
        return pupilUrl;
    }
}

File: app/src/main/java/com/example/journal/LoginManager.kt
package com.example.journal

import android.os.Build
import java.io.DataOutputStream
import java.io.FileWriter
import java.net.*
import java.nio.charset.StandardCharsets

class LoginManager(rootDirectory: String) {
    private val USER_AGENT = "Mozilla/5.0"
    private val ROOT_DIRECTORY = rootDirectory
    private val LOGIN_PRIMARY_DOMAIN = "https://schools.by/login"

    var csrftoken: String? = null
        private set
    var sessionid: String? = null
        private set
    var pupilUrl: String? = null
        private set

    @Throws(Exception::class)
    fun tryToLogin(username: String, password: String): Boolean {
        if (csrftoken == null) csrftoken = newCsrftoken()

        val postParameters = getEncodedPostParameters(username, password)

        val connection = URL(LOGIN_PRIMARY_DOMAIN).openConnection() as HttpURLConnection

        connection.connectTimeout = 1500
        connection.readTimeout = 1500

        connection.instanceFollowRedirects = false
        connection.useCaches = false
        connection.requestMethod = "POST"

        connection.setRequestProperty("accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
        connection.setRequestProperty("accept-encoding", "gzip, deflate, br")
        connection.setRequestProperty("accept-language", "en-gb")
        connection.setRequestProperty("Connection", "keep-alive")
        connection.setRequestProperty("content-length", Integer.toString(postParameters!!.length))
        connection.setRequestProperty("content-type", "application/x-www-form-urlencoded")
        connection.setRequestProperty("cookie", "csrftoken=$csrftoken;")
        connection.setRequestProperty("origin", "https://schools.by")
        connection.setRequestProperty("referer", "https://schools.by/login")
        connection.setRequestProperty("user-agent", USER_AGENT)

        connection.doInput = true
        connection.doOutput = true

        val stream = DataOutputStream(connection.outputStream)
        stream.writeBytes(postParameters)
        stream.flush()
        stream.close()

        connection.getContent()

        val responseCode = connection.responseCode
        when (responseCode) {
            302 -> {
                println("success")
            }
            200 -> {
                connection.disconnect()
                return false
            }
            else -> {
                connection.disconnect()
                throw Exception("Unknown response code")
            }
        }

        val headers = connection.headerFields
        connection.disconnect()

        for (entry in headers.entries) {
            try {
                when (entry.key) {
                    "Set-Cookie" -> {
                        for (str in entry.value) {
                            val cookies = HttpCookie.parse(str)
                            for (cookie in cookies) {
                                if (cookie.name.equals("sessionid")) sessionid = cookie.value
                            }
                        }
                    }
                    "Location" -> {
                        pupilUrl = entry.value.get(0)
                    }
                }
            } catch (npe: NullPointerException) {continue}
        }


        if (sessionid == null || pupilUrl == null) {
            throw Exception("sessionid or pupilUrl missing")
        } else {
            return true
        }
    }

    @Throws(Exception::class)
    private fun newCsrftoken(): String {
        val con = URL(LOGIN_PRIMARY_DOMAIN).openConnection() as HttpURLConnection

        con.connectTimeout = 1500
        con.readTimeout = 1500
        con.requestMethod = "GET"
        con.setRequestProperty("User-Agent", USER_AGENT)

        con.connect()
        val hf = con.headerFields
        con.disconnect()

        for (entry in hf) {
            try {
                if (entry.key == "Set-Cookie") {
                    for (str in entry.value) {
                        val cookies = HttpCookie.parse(str)
                        for (cookie in cookies) {
                            if (cookie.name == "csrftoken") {
                                return cookie.value
                            }
                        }
                    }
                }
            } catch (npe: NullPointerException) {continue}
        }
        throw Exception("csrftoken missing")
    }

    fun writeLoginDataToFiles(username: String, realName: String) {
        val foutUsername = FileWriter("$ROOT_DIRECTORY/UserData/username.txt", false)
        foutUsername.write(username)
        foutUsername.flush()
        foutUsername.close()

        val foutSessionid = FileWriter("$ROOT_DIRECTORY/UserData/sessionid.txt", false)
        foutSessionid.write(sessionid)
        foutSessionid.flush()
        foutSessionid.close()

        val foutUrl = FileWriter("$ROOT_DIRECTORY/UserData/pupilUrl.txt", false)
        foutUrl.write(pupilUrl)
        foutUrl.flush()
        foutUrl.close()

        val foutRealName = FileWriter("$ROOT_DIRECTORY/UserData/realName.txt", false)
        foutRealName.write(realName)
        foutRealName.flush()
        foutRealName.close()

        val foutStatus = FileWriter("$ROOT_DIRECTORY/UserData/status.txt", false)
        foutStatus.write("YES")
        foutStatus.flush()
        foutStatus.close()
    }

    private fun getEncodedPostParameters(username: String, password: String) : String {
        var pp = "csrfmiddlewaretoken=$csrftoken"
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            pp += "&username="
            pp += URLEncoder.encode(username, StandardCharsets.UTF_8.toString())
            pp += "&password="
            pp += URLEncoder.encode(password, StandardCharsets.UTF_8.toString())
        } else {
            pp += "&username="
            pp += URLEncoder.encode(username)
            pp += "&password="
            pp += URLEncoder.encode(password)
        }
        return pp
    }
}